diesel/src/types/mod.rs

```diff
@@ -274,10 +275,10 @@ pub type VarChar = Text;
 pub use pg::types::sql_types::*;
 
 pub trait HasSqlType<ST>: TypeMetadata {
-    fn metadata() -> Self::TypeMetadata;
+    fn metadata(lookup: &Self::MetadataLookup) -> Self::TypeMetadata;
```

Comment:
As long as one ignores the error handling this version seems to be easier. 
As soon as you add error handling you will need to touch `HasSqlType` again, because the lookup will return an `Result`. Therefore `HasSqlType::metadata` will also return an `Result`. So each Implementation and each call of `HasSqlType::metadata` will be touched again.

Using the `Dynamic/Static` approach the lookup is done inside of `ToSql` (already returns `Result`) or inside of `BindCollector::push_bound_value` (already returns `Result`). So only an `try!` around those lookup calls will be needed.